File: 2501~3000/2514. Count Anagrams/main.py
"""
Source: t.ly/VhvYg
Date: 2023/9/26
Skill:
Runtime: 33 ms, faster than 100%
Memory Usage: 16.3 MB, less than 25%
Time complexity:
Space complexity:
Constraints:

"""

import math
import sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
from functools import cache


class Solution:
    def countAnagrams(self, s: str) -> int:
        res, sz, idx, mod = 1, len(s), 0, 1e9+7
        while idx < sz:
            num, cnt = 0, [0 for _ in range(26)]
            while s[idx] != ' ':
                num += 1
                cnt[s[idx] - 'a'] += 1
                idx += 1
            res *= math.factorial(num)
            for i in range(26):
                if cnt[i] > 1:
                    res /= math.factorial(cnt[i])
            idx += 1
        return res % mod


if __name__ == "__main__":
    s = Solution()

File: 2501~3000/2514. Count Anagrams/main.cpp
/**
 * Source: https://rb.gy/hfkz8
 * Date: 2023/10/28
 * Skill:
 * Runtime: 34 ms, faster than 91.98% of C++ online submissions
 * Memory Usage: 11.72 MB, less than 91.98% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <unordered_set>

using namespace std;

#define ll long long

const int MX = 1e5+1;

class Solution {
    ll mod = 1e9+7;
public:
    int countAnagrams(string s) {
        ll res = 1, fac[MX] = {1};
        for (int i = 1; i < MX; i++) fac[i] = fac[i - 1] * i % mod;
        int len = s.length(), cnt[26];
        for (int i = 0; i < len; i++) {
            for (int j = 0; j < 26; j++) cnt[j] = 0;
            int num = 0;
            while (i < len && s[i] != ' ') {
                cnt[s[i++] - 'a']++;
                num++;
            }
            ll res1 = fac[num];
            for (int j = 0; j < 26; j++) {
                if (cnt[j] > 1) {
                    res1 = res1 * inv(fac[cnt[j]]) % mod;
                }
            }
            res = res * (res1 % mod) % mod;
        }
        return res;
    }
    ll inv(int x) {
        ll s = 1;
        for (; x > 1; x = mod % x) s = s * (mod - mod / x) % mod;
        return s;
    }
};

int main() {
    Solution s;
}